orttraining/tools/scripts/pipeline_model_split.py

```diff
@@ -0,0 +1,376 @@
+    receive_output_signal.CopyFrom(helper.make_tensor_value_info(
+        'receive_output_signal', onnx.TensorProto.BOOL, None))
+
+    new_send = model.graph.node.add()
+    new_send.CopyFrom(helper.make_node(
+        'Send',
+        inputs=['send_signal'],
+        outputs=['send_output_signal'],
+        tag=0,
+        src=0, # todo：change it when split more than 2 parts
+        dst=1,
+        domain=ms_domain,
+        element_type=1, # assuming all tensors are of type float
+        name='send'))
+
+    new_receive = model.graph.node.add()
+    new_receive.CopyFrom(helper.make_node(
+        'Recv',
+        inputs=['recv_signal'],
+        outputs=['receive_output_signal'],
+        tag=1,
+        src=0, # todo：change it when split more than 2 parts
+        dst=1,
+        domain=ms_domain,
+        element_type=1, # assuming all tensors are of type float
+        name='receive'))
+
+    new_wait = model.graph.node.add()
+    new_wait.CopyFrom(helper.make_node(
+        'WaitEvent',
+        inputs=['wait_signal'],
+        outputs=[],
+        domain=ms_domain))
+
+    new_record = model.graph.node.add()
+    new_record.CopyFrom(helper.make_node(
+        'RecordEvent',
+        inputs=['record_signal'],
+        outputs=[],
+        domain=ms_domain))
+
+    for i in range(len(upstream_nodes)):
+        n = upstream_nodes[i]
+        output_type = output_shapes[i]
+
+        output_nodes = find_all_output_nodes_by_edge(model, n.output[0])
+
+        # deal with shape inference for newly added edge
+        new_send_input_name = n.output[0] + '_send'
+        add_expand_shape(model, new_send_input_name, output_type)
+
+        new_receive_output_name = n.output[0] + '_recv'
+        add_expand_shape(model, new_receive_output_name, output_type)
+
+        new_wait_output_name = n.output[0] + '_wait'
+        add_expand_shape(model, new_wait_output_name, output_type)
+
+        # the order of data flow is: node-output -> record -> send -> recv -> wait -> node-input
+        new_record.input.extend([n.output[0]])
+        new_record.output.extend([new_send_input_name])
+
+        new_send.input.extend([new_send_input_name])
+        new_receive.output.extend([new_receive_output_name])
+
+        new_wait.input.extend([new_receive_output_name])
+        new_wait.output.extend([new_wait_output_name])
+
+        for output_node in output_nodes:
+            for i in range(len(output_node.input)):
+                for edgeId in edgeIds:
+                    if output_node.input[i] == edgeId:
+                        output_node.input[i] = new_wait_output_name
+
+    return new_send, new_receive
+
+def find_all_input_nodes(model, node):
+    nodes = []
+    inputs = []
+
+    if node:
+        for inputId in node.input:
+            for node in model.graph.node:
+                for output in node.output:
+                    if output == inputId:
+                        nodes.append(node)
```

Comment:
could be simplified to nodes.extend([n for n in model.graph.node if inputId in n.output])